File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/ClientApi/Operations/Server/Compact.cs
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Operations;
using Raven.Client.Documents.Operations.Indexes;
using Raven.Client.ServerWide;
using Raven.Client.ServerWide.Operations;

namespace Raven.Documentation.Samples.ClientApi.Operations.Server
{
    public class Compact
    {
        private interface IFoo
        {
            /*
            #region syntax_1
            public CompactDatabaseOperation(CompactSettings compactSettings)
            #endregion
            */
        }

        public Compact()
        {
            using (var documentStore = new DocumentStore())
            {
                #region compact_0
                // Define the compact settings
                CompactSettings settings = new CompactSettings
                {
                    // Database to compact
                    DatabaseName = "Northwind",

                    // Set 'Documents' to true to compact all documents in database
                    // Indexes are not set and will not be compacted
                    Documents = true
                };
                
                // Define the compact operation, pass the settings
                IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);
                
                // Execute compaction by passing the operation to Maintenance.Server.Send
                Operation operation = documentStore.Maintenance.Server.Send(compactOp);
                // Wait for operation to complete
                operation.WaitForCompletion();
                #endregion
            }
            
            using (var documentStore = new DocumentStore())
            {
                #region compact_1
                // Define the compact settings
                CompactSettings settings = new CompactSettings
                {
                    // Database to compact
                    DatabaseName = "Northwind",
                    
                    // Setting 'Documents' to false will compact only the specified indexes
                    Documents = false,
                    
                    // Specify which indexes to compact
                    Indexes = new[] { "Orders/Totals", "Orders/ByCompany" }, 
                    
                    // Optimize indexes is Lucene's feature to gain disk space and efficiency
                    // Set whether to skip this optimization when compacting the indexes
                    SkipOptimizeIndexes = false
                };
                
                // Define the compact operation, pass the settings
                IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);

                // Execute compaction by passing the operation to Maintenance.Server.Send
                Operation operation = documentStore.Maintenance.Server.Send(compactOp);
                // Wait for operation to complete
                operation.WaitForCompletion();
                #endregion
            }
            
            using (var documentStore = new DocumentStore())
            {
                #region compact_2
                // Get all indexes names in the database using the 'GetIndexNamesOperation' operation
                // Use 'ForDatabase' if the target database is different than the default database defined on the store
                string[] allIndexNames =
                    documentStore.Maintenance.ForDatabase("Northwind")
                        .Send(new GetIndexNamesOperation(0, int.MaxValue));
                
                // Define the compact settings
                CompactSettings settings = new CompactSettings
                {
                    DatabaseName = "Northwind", // Database to compact
                    
                    Documents = true,           // Compact all documents

                    Indexes = allIndexNames,    // All indexes will be compacted
                    
                    SkipOptimizeIndexes = true  // Skip Lucene's indexes optimization
                };
                
                // Define the compact operation, pass the settings
                IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);
                
                // Execute compaction by passing the operation to Maintenance.Server.Send
                Operation operation = documentStore.Maintenance.Server.Send(compactOp);
                // Wait for operation to complete
                operation.WaitForCompletion();
                #endregion
            }
            
            using (var documentStore = new DocumentStore())
            {
                #region compact_3
                // Get all member nodes in the database-group using the 'GetDatabaseRecordOperation' operation
                var allMemberNodes =
                    documentStore.Maintenance.Server.Send(new GetDatabaseRecordOperation("Northwind"))
                        .Topology.Members;

                // Define the compact settings as needed
                CompactSettings settings = new CompactSettings
                {
                    // Database to compact
                    DatabaseName = "Northwind",
                    
                    //Compact all documents in database
                    Documents = true
                };

                // Execute the compact operation on each member node
                foreach (string nodeTag in allMemberNodes)
                {
                    // Define the compact operation, pass the settings
                    IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);
                    
                    // Execute the operation
                    // Use `ForNode` to specify the node to operate on
                    Operation operation = documentStore.Maintenance.Server.ForNode(nodeTag).Send(compactOp);
                    // Wait for operation to complete
                    operation.WaitForCompletion();
                }
                #endregion
            }
        }

        public async Task CompactAsync()
        {
            using (var documentStore = new DocumentStore())
            {
                #region compact_0_async
                // Define the compact settings
                CompactSettings settings = new CompactSettings
                {
                    // Database to compact
                    DatabaseName = "Northwind",

                    // Set 'Documents' to true to compact all documents in database
                    // Indexes are not set and will not be compacted
                    Documents = true
                };
                
                // Define the compact operation, pass the settings
                IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);
                
                // Execute compaction by passing the operation to Maintenance.Server.SendAsync
                Operation operation = await documentStore.Maintenance.Server.SendAsync(compactOp);
                // Wait for operation to complete
                await operation.WaitForCompletionAsync().ConfigureAwait(false);
                #endregion
            }
            
            using (var documentStore = new DocumentStore())
            {
                #region compact_1_async
                // Define the compact settings
                CompactSettings settings = new CompactSettings
                {
                    // Database to compact
                    DatabaseName = "Northwind",
                    
                    // Setting 'Documents' to false will compact only the specified indexes
                    Documents = false,
                    
                    // Specify which indexes to compact
                    Indexes = new[] { "Orders/Totals", "Orders/ByCompany" }, 
                    
                    // Optimize indexes is Lucene's feature to gain disk space and efficiency
                    // Set whether to skip this optimization when compacting the indexes
                    SkipOptimizeIndexes = false
                };
                
                // Define the compact operation, pass the settings
                IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);

                // Execute compaction by passing the operation to Maintenance.Server.SendAsync
                Operation operation = await documentStore.Maintenance.Server.SendAsync(compactOp);
                // Wait for operation to complete
                await operation.WaitForCompletionAsync().ConfigureAwait(false);
                #endregion
            }
            
            using (var documentStore = new DocumentStore())
            {
                #region compact_2_async
                // Get all indexes names in the database using the 'GetIndexNamesOperation' operation
                // Use 'ForDatabase' if the target database is different than the default database defined on the store
                string[] allIndexNames =
                    documentStore.Maintenance.ForDatabase("Northwind")
                        .Send(new GetIndexNamesOperation(0, int.MaxValue));
                
                // Define the compact settings
                CompactSettings settings = new CompactSettings
                {
                    DatabaseName = "Northwind", // Database to compact
                    
                    Documents = true,           // Compact all documents

                    Indexes = allIndexNames,    // All indexes will be compacted
                    
                    SkipOptimizeIndexes = true  // Skip Lucene's indexes optimization
                };
                
                // Define the compact operation, pass the settings
                IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);
                
                // Execute compaction by passing the operation to Maintenance.Server.SendAsync
                Operation operation = await documentStore.Maintenance.Server.SendAsync(compactOp);
                // Wait for operation to complete
                await operation.WaitForCompletionAsync();
                #endregion
            }
            
            using (var documentStore = new DocumentStore())
            {
                #region compact_3_async
                // Get all member nodes in the database-group using the 'GetDatabaseRecordOperation' operation
                var allMemberNodes =
                    documentStore.Maintenance.Server.Send(new GetDatabaseRecordOperation("Northwind"))
                        .Topology.Members;

                // Define the compact settings as needed
                CompactSettings settings = new CompactSettings
                {
                    // Database to compact
                    DatabaseName = "Northwind",
                    
                    //Compact all documents in database
                    Documents = true
                };

                // Execute the compact operation on each member node
                foreach (string nodeTag in allMemberNodes)
                {
                    // Define the compact operation, pass the settings
                    IServerOperation<OperationIdResult> compactOp = new CompactDatabaseOperation(settings);
                    
                    // Execute the operation
                    // Use `ForNode` to specify the node to operate on
                    Operation operation = await documentStore.Maintenance.Server.ForNode(nodeTag).SendAsync(compactOp);
                    // Wait for operation to complete
                    await operation.WaitForCompletionAsync();
                }
                #endregion
            }
        }
    }
}

File: Documentation/5.4/Samples/nodejs/client-api/operations/Server/compact.js
import { DocumentStore } from "ravendb";

const documentStore = new DocumentStore();

async function compact() {
    {
        //region compact_0
        // Define the compact settings
        const compactSettings = {
            // Database to compact
            databaseName: "Northwind",

            // Set 'documents' to true to compact all documents in database
            // Indexes are not set and will not be compacted
            documents: true
        };

        // Define the compact operation, pass the settings
        const compactOp = new CompactDatabaseOperation(compactSettings);

        // Execute compaction by passing the operation to maintenance.server.send
        const asyncOperation = await documentStore.maintenance.server.send(compactOp);
        
        // Wait for operation to complete, during compaction the database is offline
        await asyncOperation.waitForCompletion();
        //endregion    
    }
    {
        //region compact_1
        // Define the compact settings
        const compactSettings = {
            // Database to compact
            databaseName: "Northwind",

            // Setting 'documents' to false will compact only the specified indexes
            documents: false,

            // Specify which indexes to compact
            indexes: ["Orders/Totals", "Orders/ByCompany"] 
        };

        // Define the compact operation, pass the settings
        const compactOp = new CompactDatabaseOperation(compactSettings);

        // Execute compaction by passing the operation to maintenance.server.send
        const asyncOperation = await documentStore.maintenance.server.send(compactOp);
        // Wait for operation to complete
        await asyncOperation.waitForCompletion();
        //endregion    
    }
    {
        //region compact_2
        // Get all indexes names in the database using the 'GetIndexNamesOperation' operation
        // Use 'forDatabase' if the target database is different than the default database defined on the store
        const allIndexNames = await documentStore.maintenance.forDatabase("Northwind")
            .send(new GetIndexNamesOperation(0, 50));

        // Define the compact settings
        const compactSettings = {
            databaseName: "Northwind", // Database to compact

            documents: true,           // Compact all documents

            indexes: allIndexNames,    // All indexes will be compacted
        };

        // Define the compact operation, pass the settings
        const compactOp = new CompactDatabaseOperation(settings);
    
        // Execute compaction by passing the operation to maintenance.server.send
        const asyncOperation = await documentStore.maintenance.server.send(compactOp);
        // Wait for operation to complete
        await asyncOperation.waitForCompletion();
        //endregion    
    }
    {
        //region compact_3
        // Get all member nodes in the database-group using the 'GetDatabaseRecordOperation' operation
        const databaseRecord =
            await documentStore.maintenance.server.send(new GetDatabaseRecordOperation("Northwind"));
        const allMemberNodes = databaseRecord.topology.members;

        // Define the compact settings as needed
        const compactSettings = {
            // Database to compact
            databaseName: "Northwind",

            //Compact all documents in database
            documents: true
        };

        // Execute the compact operation on each member node
        for (let i = 0; i < allMemberNodes.length; i++) {
            // Define the compact operation, pass the settings
            const compactOp = new CompactDatabaseOperation(compactSettings);
            
            // Execute the operation on a specific node
            // Use `forNode` to specify the node to operate on
            const serverOpExecutor = await documentStore.maintenance.server.forNode(allMemberNodes[i]);
            const asyncOperation = await serverOpExecutor.send(compactOp);
            
            // Wait for operation to complete
            await asyncOperation.waitForCompletion();
        }
        //endregion
    }
}

{
    //region syntax
    const compactOperation = new CompactDatabaseOperation(compactSettings);
    //endregion
}
